### PR TITLE
Add support for v3.13

### PR DESCRIPTION
**Pull request by NickCondron**
Decided to add placement as a field of a `PlayerEnd` struct analogous to the `Player` struct in `Start`. I think this way is a little nicer to use, and there's a good chance more player-specific fields are added in the future. Maybe we could choose better names.

### CODE REVIEW

**Review by hohav — COMMENTED** [review]
I like this design. Could you add a v3.13 test?

File: peppi/src/model/game.rs [hunk 1]
```diff
@@ -278,6 +289,7 @@ pub struct GeckoCodes {
 /// Replay data for a single game of Melee.
 ///
 /// See https://github.com/project-slippi/slippi-wiki/blob/master/SPEC.md.
+#[non_exhaustive]
```
Comment (hohav) [inline peppi/src/model/game.rs:292]: This is unrelated AFAICT, and should go in a separate PR.

But I also haven't decided how I feel about `non_exhastive`. I don't expect callers to construct their own game objects, but on the other hand they should be able to do that if they want— I just don't care if their code then breaks when we add new fields. And I'm not interested in forcing constructors everywhere just for this use case. 🤷

Comment (NickCondron) [reply peppi/src/model/game.rs:292]: This was added by accident oops. FWIW I was planning on making a PR adding `non_exhaustive` in different places. In my opinion the main benefit is reducing the amount of changes that would be considered breaking.

For example, when you added a new field `debug_dir` to `serde::de::Opts` my code broke. If `Opts` was `non_exhaustive` I could write something like:

```rust
let opts = Opts { skip_frames: true, ..Default::default() };
```

Then new settings don't break the API (in this case we would need to maintain a default impl, but that is trivial).

File: peppi/src/serde/de.rs [hunk 2]
```diff
@@ -438,13 +438,37 @@ pub(crate) fn game_start(r: &mut &[u8]) -> Result<game::Start> {
 	})
 }
 
+pub fn player_end(port: Port, placement: i8) -> Result<Option<game::PlayerEnd>> {
+	match placement {
+		-1 => Ok(None),
+		0..=3 => Ok(Some(game::PlayerEnd { port, placement: placement as u8 })),
```
[the remark this answers is not in the document]
Comment (NickCondron) [reply peppi/src/serde/de.rs:444]: patterns can't be exclusive ranges currently https://doc.rust-lang.org/beta/unstable-book/language-features/exclusive-range-pattern.html

I can add a line like `const MAX_PLACEMENT: i8 = NUM_PORTS as i8 - 1;` and use that, but idk how much that helps readability.

File: peppi/src/serde/de.rs [hunk 3]
```diff
@@ -438,13 +438,35 @@ pub(crate) fn game_start(r: &mut &[u8]) -> Result<game::Start> {
 	})
 }
 
+pub fn player_end(port: Port, placement: i8) -> Result<Option<game::PlayerEnd>> {
+	match placement {
+		-1 => Ok(None),
+		0..=MAX_PLACEMENT => Ok(Some(game::PlayerEnd { port, placement: placement as u8, })),
+		p => Err(err!("Invalid player placement {}", p)),
```
Comment (hohav) [inline peppi/src/serde/de.rs:445]: I don't think we need to reject out-of-range placement values. If I were parsing a replay and the only thing "wrong" with it was that someone got a placement of 99 (maybe due to some Melee or Slippi bug), I'd prefer to get the value as written in the file. And since we store the placement as a u8, in theory users of peppi need to guard against high values anyway (unless we create a type for placements that can never be above 3, but I don't think that's worth it).

File: peppi/src/model/game.rs [hunk 4]
```diff
@@ -172,19 +173,26 @@ pseudo_enum!(EndMethod: u8 {
...
 /// Information about the end of the game.
 #[derive(Clone, Debug, PartialEq, Eq, Serialize)]
 pub struct End {
 	/// how the game ended
 	pub method: EndMethod,
 
-	/// mostly-redundant copy of the raw start block, for round-tripping
-	#[serde(skip)] #[doc(hidden)]
-	pub raw_bytes: Vec<u8>,
-
```
Comment (hohav) [inline peppi/src/model/game.rs:184]: peppi-slp needs this for [easy round-tripping](https://github.com/hohav/peppi-slp/blob/8eaa272d1f1a3f281c3a085d1aab8b00c104f9f0/src/main.rs#L110). Yes, for game end we could reconstruct the original bytes fairly easily, but for an event that happens only once I value concise and bullet-proof code over a few extra bytes.

(Although the docstring should say "end event" rather than "start block", and the corresponding comment in `Start` should say "start event".)

I should really create some integration tests between peppi-slp and peppi, I just haven't figured out how to do that as part of peppi's CI.

File: peppi/src/model/game.rs [hunk 5]
```diff
@@ -172,19 +173,26 @@ pseudo_enum!(EndMethod: u8 {
...
 	/// player who LRAS'd, if any (added: v2.0)
 	#[serde(skip_serializing_if = "Option::is_none")]
 	pub lras_initiator: Option<Option<Port>>,
+
+	/// player-specific data (added v3.13)
```
Comment (hohav) [inline peppi/src/model/game.rs:193]: nit: please add a colon after "added" 🙏

File: peppi/src/model/game.rs (hunk 1 above)
Comment (hohav) [reply peppi/src/model/game.rs]: That [still doesn't work](https://users.rust-lang.org/t/using-default-default-in-a-non-exhaustive-struct-fails/51193) though, right? Although I agree with adding `Default` to `Opts`.

File: peppi/src/serde/de.rs (hunk 3 above)
Comment (NickCondron) [reply peppi/src/serde/de.rs:445]: Maybe a type for placements is worth it? This is not too bad:

```rust
pseudo_enum!(Placement: u8) {
  0 => FIRST,
  1 => SECOND,
  2 => THIRD,
  3 => FOURTH
}
```

-------

I agree you can't guarantee melee/slippi is correct, but I see only a few options:

1. Error when file doesn't match the spec (my original solution)
2. Provide complete transparency to the user -- just give them the `i8` that's in the file because we can't guarantee that it's always [-1,3]
3. Perform some destructive mapping (your proposed solution) -- negative values in the file would probably have to map to `None`

Solution 2 seems defeatist and unhelpful to users in 99.9% of scenarios (suppose <0.1% of the time there's a wrong value and it's helpful to expose the mistake to users). We already fail in other places where a value doesn't match the spec (eg. l-cancel in post frame update).

Comment (hohav) [reply peppi/src/serde/de.rs]: I don't think an enum carries any value for a "placement", especially since most people writing code will be concerned with 2p games and so will only be handling a subset of the possible values anyway. But it should be possible to write generic code that cares about placement ordering, and a `u8` is perfect for that.

My original objection was informed by having written strict parsing code that rejected "unexpected" enum values, and then failing to parse real games because there were undocumented action states etc. That's why `pseudo_enum` is a thing. But I suppose that's less likely for placements than for action states, and would possibly have worse consequences if it were caused by a Slippi bug. So I think I'm fine with keeping the original assertion after all.

**Comment by NickCondron** [comment]
Rebased. Test are failing because of a bug I'm looking into. It is unrelated to these v3.13 changes and I believe it is caused by issues with parsing the gecko codes & event splitter events.

**Comment by hohav** [comment]
Merged via 6cd2ef47417cc02476eb845e2548d8043793090c.
